### PR TITLE
UDP: Ensure propagation of control message to pool.Message

### PR DESCRIPTION
**Pull request by jkralik**
This commit enhances the UDP functionality, ensuring proper dissemination of control messages to pool.Message for improved network coordination and responsiveness

### CODE REVIEW

File: net/connUDP.go [hunk 1]
```diff
@@ -24,10 +24,36 @@ type UDPConn struct {
 }
 
 type ControlMessage struct {
+	Dst     net.IP // destination address, receiving only
 	Src     net.IP // source address, specifying only
```
Comment (HRogge) [inline net/connUDP.go:28]: Source-Address is also valid when receiving packet, same address as the "remote" parameter you get from the read

File: net/connUDP.go [hunk 2]
```diff
@@ -24,10 +24,36 @@ type UDPConn struct {
...
 	IfIndex int    // interface index, must be 1 <= value when specifying
```
Comment (HRogge) [inline net/connUDP.go:29]: Interface-Index 0 is defined as "any interface", so I think the "must be 1<=" part of the comment can be removed.

**Review by HRogge — APPROVED** [review]
I just checked out the latest commit of this branch and I successfully got src/dst-IP address and Interface Index. I am looking forward to see this merged.

File: net/connUDP.go (hunk 1 above)
Comment (jkralik) [reply net/connUDP.go:28]: According to doc the https://pkg.go.dev/golang.org/x/net@v0.18.0/ipv4#ControlMessage it seems that `Src` cannot be set for write.

Comment (HRogge) [reply net/connUDP.go:28]: doc says "specifying only" for source (as in "write only).

but I am quite sure (just tested it on Debian) that you ALSO get the source set when receiving.

**Comment by HRogge** [comment]
Just re-tested with the current commit, still works fine.

File: net/connUDP.go (hunk 2 above)
Comment (jkralik) [reply net/connUDP.go]: Fixed, thx :)

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -30,34 +30,71 @@ jobs:
...
+      - name: Download artifacts
+        uses: actions/download-artifact@v3
+        with:
+          path: ./outputs
+
+      - name: Prepare coverage and report files
+        run: |
+          mkdir -p .tmp/coverage
+          mkdir -p .tmp/report
+          find ./outputs -name "*.coverage.txt" -exec sh -c 'cp $1 .tmp/coverage/$(echo $1 | sed "s/[\/.]/-/g" ).coverage.txt' _ {}  \;
```
Comment (Danielius1922) [inline .github/workflows/test.yml:88]: From the job logs, the file name are kinda weird ("--outputs" prefix and "-coverage-txt.coverage.txt" suffix)

`/github/workspace/.tmp/coverage/--outputs-608abf8a79c6238b98b3e145b34057c18dd699466c788dc7ff714116d607577d-ubuntu-latest-coverage-txt.coverage.txt`

but if the coverage calculation works then who cares. :D

File: net/connUDP.go [hunk 4]
```diff
@@ -408,8 +513,18 @@ func (c *UDPConn) writeMulticast(ctx context.Context, raddr *net.UDPAddr, buffer
...
+		// because the connection is already established.
+		// Note: Overwriting the destination address is only supported on Linux.
+		return c.connection.Write(buffer)
+	}
+	return c.packetConn.WriteTo(buffer, cm, raddr)
+}
+
 // WriteWithContext writes data with context.
-func (c *UDPConn) WriteWithContext(ctx context.Context, raddr *net.UDPAddr, buffer []byte) error {
+func (c *UDPConn) WriteWithContext(ctx context.Context, raddr *net.UDPAddr, cm *ControlMessage, buffer []byte) error {
```
Comment (Danielius1922) [inline net/connUDP.go:527]: Is changing signatures `WriteWithContext` and `ReadWithContext` unproblematic although they are exported?

Comment (HRogge) [reply net/connUDP.go:527]: It might break a few applications... easy to repair, but unexpected.

**Review by Danielius1922 — APPROVED** [review]

**Comment by jkralik** [comment]
@HRogge  Pls this is a final state. Could you check if it works as is expected for you ?

**Comment by HRogge** [comment]
> @HRogge Pls this is a final state. Could you check if it works as it is expected ?

It seems that multicast transmission is broken... not sure why, still trying to understand. I get neither a compiler nor a runtime error when using WriteMulticast() on net.UDPConn

**Comment by HRogge** [comment]
I also get a broken Date for the current commit when I look at it with 'git log':
```
commit 7e6b2bbd34f0051d0ff09d06af2d453450a30706 (HEAD -> jkralik/feature/udp-propagate-interface, github-upstream/jkralik/feature/udp-propagate-interface)
Author: Jozef Kralik <jojo.lwin@gmail.com>
Date:   Wed May 10 15:11:10 2023 +0000

    Update UDP: Ensure propagation of control message to pool.Message
    
    This commit enhances the UDP functionality, ensuring proper dissemination
    of control messages to pool.Message for improved network coordination
    and responsiveness
```

**Comment by HRogge** [comment]
> > @HRogge Pls this is a final state. Could you check if it works as it is expected ?
> 
> It seems that multicast transmission is broken... not sure why, still trying to understand. I get neither a compiler nor a runtime error when using WriteMulticast() on net.UDPConn

It seems my multicast now ignores the specified interface and goes along the default route (into my local management interface).

**Comment by HRogge** [comment]
> > > @HRogge Pls this is a final state. Could you check if it works as it is expected ?
> > 
> > 
> > It seems that multicast transmission is broken... not sure why, still trying to understand. I get neither a compiler nor a runtime error when using WriteMulticast() on net.UDPConn
> 
> It seems my multicast now ignores the specified interface and goes along the default route (into my local management interface).

***facepalm*** okay, got the problem resolved... stupidity on my side.

**Review by HRogge — APPROVED** [review]
Code works fine with the multicast application I have been using for testing. Looking forward to see it merged.

**Pull request merged**